service/src/main/java/com/codeforcommunity/requester/TreeBenefitsCalculator.java

```diff
@@ -0,0 +1,189 @@
+import static org.jooq.generated.Tables.TREE_SPECIES;
+
+import com.codeforcommunity.exceptions.ResourceDoesNotExistException;
+import java.util.HashMap;
+import java.util.Map;
+import org.jooq.DSLContext;
+import org.jooq.Field;
+import org.jooq.generated.tables.records.TreeSpeciesRecord;
+
+public class TreeBenefitsCalculator {
+  private final DSLContext db;
+  final String speciesCode;
+  final double diameterCM;
+  static final double[] intervals =
+      new double[] {3.81, 11.43, 22.86, 38.10, 53.34, 68.58, 83.82, 99.06, 114.30};
+  final double x0;
+  final double x1;
+  static final Map<String, Double> currencyConversion =
+      new HashMap<String, Double>() {
+        {
+          put("electricity_kwh_to_currency", 0.1401);
+          put("natural_gas_kbtu_to_currency", 0.01408);
+          put("h20_gal_to_currency", 0.0008);
+          put("co2_lb_to_currency", 0.00334);
+          put("o3_lb_to_currency", 4.59);
+          put("nox_lb_to_currency", 4.59);
+          put("pm10_lb_to_currency", 8.31);
+          put("sox_lb_to_currency", 3.48);
+          put("voc_lb_to_currency", 2.31);
+        }
+      };
+  static final Map<String, Double> unitConversion =
+          new HashMap<String, Double>() {
+            {
+              put("kBTU_to_kWh", 0.2930);
+              put("kg_to_lb", 2.20462);
+              put("m3_to_gal", 264.172);
+            }
+          };
+
+  // diameter in inches
+  public TreeBenefitsCalculator(DSLContext db, String commonName, double diameter) {
+    this.db = db;
+    this.speciesCode = getSpeciesCode(commonName);
+    this.diameterCM = diameter * 2.54;
+
+    double x0 = 0;
+    double x1 = 0;
+    for (int i = 0; i < intervals.length; i++) {
+      if (intervals[i] <= diameterCM) {
+        x0 = intervals[i];
+        x1 = intervals[i + 1];
+      }
+    }
+    if (x0 == 0 || x1 == 0) {
+      throw new IllegalArgumentException("diameter must be non-zero");
+    }
+    this.x0 = x0;
+    this.x1 = x1;
+  }
+
+  // helper to query the species code from the tree_species table
+  private String getSpeciesCode(String commonName) {
+    TreeSpeciesRecord record =
+        db.selectFrom(TREE_SPECIES).where(TREE_SPECIES.COMMON_NAME.eq(commonName)).fetchOne();
+    if (record == null) {
+      throw new IllegalArgumentException("Unable to find entry with name " + commonName);
+    }
+    String speciesCode = record.getSpeciesCode();
+    return speciesCode;
+  }
+
+  // helper to calculate the interpolated value of the given property
+  private double calcProperty(Field property) {
+    double y0 =
+        db.selectFrom(TREE_BENEFITS)
+            .where(TREE_BENEFITS.SPECIES_CODE.eq(speciesCode))
+            .and(TREE_BENEFITS.DIAMETER.eq(x0))
+            .fetchOne(property, Double.class);
+    double y1 =
+        db.selectFrom(TREE_BENEFITS)
+            .where(TREE_BENEFITS.SPECIES_CODE.eq(speciesCode))
+            .and(TREE_BENEFITS.DIAMETER.eq(x1))
+            .fetchOne(property, Double.class);
```

Comment:
⛏️ Instead of querying our database each time we call `calcProperty()`, could we instead fetch once in the constructor, store those records as properties, and fetch the given field from those records in this method? That way, we won't have to keep fetching the same information multiple times. The signature won't need to change, since we can call `get(Field property)` on a database record.

Reply:
you are right-- I refactored